server/parsec/components/auth.py

```diff
@@ -155,6 +162,77 @@ def verify_signature(self, verify_key: VerifyKey) -> bool:
             return False
 
 
+class AccountAuthenticationType(enum.Enum):
+    Password = b"password"
+
+
+@dataclass
+class AccountAuthenticationToken:
+    auth_type: AccountAuthenticationType
+    timestamp: DateTime
+    email: EmailStr
+    header_and_payload: tuple[bytes, ...]
+    signature: bytes
+
+    HEADER = b"PARSEC-HMAC-BLAKE2B"
+
+    @staticmethod
+    def generate_raw(
+        auth_type: AccountAuthenticationType,
+        timestamp: DateTime,
+        email: EmailStr,
+        body: bytes,
+        key: SecretKey,
+    ) -> bytes:
+        raw_timestamp_us = str(timestamp.as_timestamp_seconds()).encode("ascii")
+        raw_email = urlsafe_b64encode(email.encode())
+        header_and_payload = (
+            AccountAuthenticationToken.HEADER,
+            auth_type.value,
+            raw_email,
+            raw_timestamp_us,
+        )
+        body_sha256 = hashlib.sha256(body).digest()
+        raw_signature = key.hmac_full(b"".join(header_and_payload) + body_sha256)
+        return b"%s.%s" % (b".".join(header_and_payload), urlsafe_b64encode(raw_signature))
+
+    @classmethod
+    def from_raw(cls, raw: bytes) -> "AccountAuthenticationToken":
+        try:
+            raw_header_and_payload, raw_signature = raw.rsplit(b".", 1)
+            header_and_payload = raw_header_and_payload.split(b".")
+            header, raw_auth_type, raw_email, raw_timestamp_us = header_and_payload
+            if header != cls.HEADER:
+                raise ValueError
+            auth_type = AccountAuthenticationType(raw_auth_type)
+            email = urlsafe_b64decode(raw_email).decode()
+            validate_email(email)
+            timestamp = DateTime.from_timestamp_seconds(int(raw_timestamp_us))
+            signature = urlsafe_b64decode(raw_signature)
+        except (
+            ValueError,
+            # Raised when validating email
+            PydanticCustomError,
+        ) as e:
+            raise ValueError("Invalid token format") from e
```

Comment:
Maybe this would be confusing if the error is raised when the source is the email format.

Reply:
You prefer that the implementation return a dedicated error message for the email error?